doc/wavetable-communication.md

```diff
@@ -0,0 +1,184 @@
+# Wavetable communication
+
+This explains how RT and non-RT threads communicate when new waves for the
+wavetable need to be generated.
+
+## Wavetable layout
+
+A Tensor3 (3D-Tensor) has the following contents:
+
+```
+  ooooooo 1D-Tensor (frequeny values in Hz)      --
+                                                  |
+  ooooooo 1D-Tensor (semantic values, e.g.        | "wavetable scales"
+                     random seeds or              |
+                     basewave function params)   --
+
+  ooooooo 1D-Tensor (float buffer = 1 wave)      --
+  ^                                               |
+  |                                               | "wavetable data"
+  o...ooo 2D Tensor (Tensor1s for each freq)      | or just
+  ^       with ringbuffer semantics (see below)   | "waves"
+  |       (MW produces, AD note consumes)         |
+  |                                               |
+  o...ooo 3D Tensor (Tensor2s for each semantic) --
+
+```
+
+Ringbuffer layout:
+
+```
+  no required space holder (r=w is unambiguous internally)             wrap
```

Comment:
Ah, this is the point I was thinking about 30 minutes ago or so. r=w either implies writable_space=0 items or writable_space=array length items. Mind pointing me back to what eliminated this ambiguity? Other designs lost a single item from array_length items in order to prevent this issue.

Reply:
I think this is mostly described in the class comment of `AbstractRingBuffer`. The internal `r/w/w_delayed` pointers have twice the external range, so the ringbuffer can still differ between full and empty, but from outside the classes, there is a `mod` to get the read space etc.

As a reason, I found this in a very old commit 6dd7e6f2b2724193ff44679b8f2247139c72152d from Feb 2021 on a side branch:

```
    Allow w going up to r in AbstractRingbuffer
    
    Previously, it only goes up to r-1. This commit itself has no real
    functional sense, but it prepares for the following commit, which
    reduces ringbuffers to size 1. Without this commit, ringbuffers of size
    1 could never write the first element.
```

Might be good to note this somewhere if the reason is nowhere explained in the docs/code.

Reply:
Edit: Added a note in the `AbstractRingBuffer` doxygen comment.